### PR TITLE
Random orthonormal basis in matrix_props

### LINKED ISSUE #435: Feature: Generate random orthonormal basis
This particular feature may be of use if we wish to generate a collection of quantum states (basis of vectors) that have the property of being orthonormal. The columns of a randomly generated unitary matrix can be used to generate an orthonormal basis

Something like this could be used:

```py
def random_orthonormal_basis(dim: int) -> list[np.ndarray]:
    # Generate a random unitary matrix
    U = unitary_group.rvs(dim)

    # The columns of U form an orthonormal basis
    basis = [U[:, i] for i in range(dim)]
    return basis
```

This feature should live in `rand/random_orthonormal_basis.py` and would need to have the requisite coverage and testing in `tests/test_random_orthonormal_basis.py`.

### PR DESCRIPTION
**Pull request by purva-thakre**
## Description
Fixes #435 

## Changes
Notable changes that this PR has either accomplished or will accomplish. Feel free to add more lines to the itemized list
below.

  -  [ ] Change 1

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [ ] Use `ruff` for errors related to code style and formatting.
  -  [ ] Verify all previous and newly added unit tests pass in `pytest`.
  -  [ ] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [ ] Use `linkcheck` to check for broken links in the documentation
  -  [ ] Use `doctest` to verify the examples in the function docstrings work as expected.

### CODE REVIEW

File: toqito/rand/random_orthonormal_basis.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        Number of elements in the random orthonormal basis.
+
+    """
+    random_mat = random_unitary(dim, is_real=True)
+
+    rand_orth_basis = [random_mat[:, i] for i in range(dim)]
+
+    if not is_orthonormal(np.array(rand_orth_basis)):
+        raise ValueError("Random orthonormal basis could not constructed.")
+
```
Comment (purva-thakre) [inline toqito/rand/random_orthonormal_basis.py:40]: when `is_real = False`, perhaps `is_orthonormal` requires a tolerance?

**Comment by purva-thakre** [comment]
@vprusso As discussed, here's the minimum working example when `is_real=False` in `random_unitary`. The created basis elements are mutually orthogonal but they fail at the second check. 

`np.dot` is not creating a matrix close to `np.eye`. I don't think `.T` works that well when the matrix elements are complex. 

```python
import numpy as np
from toqito.rand import random_unitary
from toqito.matrix_props import is_orthonormal, is_unitary
from toqito.state_props import is_mutually_orthogonal

dim = 4
random_mat = random_unitary(dim, is_real=False)

if is_unitary(random_mat):
    print("Random matrix is unitary.")

rand_orth_basis = [random_mat[:, i] for i in range(dim)]

if is_mutually_orthogonal(rand_orth_basis):
    print("The generated basis elements are mutually orthogonal.")

if not is_orthonormal(np.array(rand_orth_basis)):
    raise ValueError("Random orthonormal basis could not constructed.")
```

File: toqito/rand/random_orthonormal_basis.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    .. bibliography::
+        :filter: docname in docnames
+
+    dim: int
+        Number of elements in the random orthonormal basis.
+
+    """
+    random_mat = random_unitary(dim, is_real)
+
+    rand_orth_basis = [random_mat[:, i] for i in range(dim)]
```
Comment (vprusso) [inline toqito/rand/random_orthonormal_basis.py:36]:
```suggestion
    random_mat = random_unitary(dim, is_real)
    return [random_mat[:, i] for i in range(dim)]
```

**Review by vprusso — APPROVED** [review]
Minor comment, but aside from that approved! LGTM!

**Pull request merged**